plugin/transfer/setup.go

```diff
@@ -0,0 +1,114 @@
+package transfer
+
+import (
+	"net"
+
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+
+	"github.com/caddyserver/caddy"
+)
+
+func init() {
+	caddy.RegisterPlugin(pluginName, caddy.Plugin{
+		ServerType: "dns",
+		Action:     setup,
+	})
+}
+
+func setup(c *caddy.Controller) error {
+	t, err := parse(c)
+
+	if err != nil {
+		return plugin.Error(pluginName, err)
+	}
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		t.Next = next
+		return t
+	})
+
+	c.OnStartup(func() error {
+		// find all plugins that implement Transferer and add them to t.Transferers
+		plugins := dnsserver.GetConfig(c).Handlers()
+		for _, pl := range plugins {
+			if tr, ok := pl.(Transferer); ok {
+				for _, z := range tr.Authoritative() {
+					for _, x := range t.xfrs {
+						var found bool
+						for _, xz := range x.Zones {
+							if z == xz {
+								found = true
+								break
+							}
+						}
+						if !found {
+							continue
+						}
+						if x.Transferers[z] != nil {
+							// Acknowledge only the first transferer for each zone.
+							// This doesn't play nice with fallthrough, where more than one plugin can
```

Comment:
Would be nice to see if a later version could do something for this case.

Reply:
Yes, I thought about this and decided it was complicated enough to leave out for now, to keep this PR small and easily reviewable.  I think it's solvable.  The transfer plugin should be able to determine the "order" of plugins and merge them down with the records of other plugins later in the chain, preferring records from earlier plugins when there are collisions.  Multi instance plugins would need be handled by each plugin's implementation of transfer.